bitcoin/src/taproot.rs

```diff
@@ -524,116 +579,158 @@ impl Default for TaprootBuilder {
-        self.0.branch()[0].as_ref().expect("from_inner only parses is_complete builders")
+        &self.0
     }
 
-    /// Converts self into builder [`TaprootBuilder`]. The builder is guaranteed to be finalized.
-    pub fn into_builder(self) -> TaprootBuilder { self.0 }
-
-    /// Constructs [`TaprootBuilder`] by internally cloning the `self`. The builder is guaranteed
-    /// to be finalized.
-    pub fn to_builder(&self) -> TaprootBuilder { self.0.clone() }
+    /// Gets the inner [`NodeInfo`] of this tree root.
+    pub fn into_node_info(self) -> NodeInfo {
+        self.0
+    }
 
     /// Returns [`TapTreeIter<'_>`] iterator for a taproot script tree, operating in DFS order over
     /// tree [`ScriptLeaf`]s.
     pub fn script_leaves(&self) -> TapTreeIter {
-        match (self.0.branch().len(), self.0.branch().last()) {
-            (1, Some(Some(root))) => TapTreeIter { leaf_iter: root.leaves.iter() },
-            // This should be unreachable as we Taptree is already finalized
-            _ => unreachable!("non-finalized tree builder inside TapTree"),
-        }
+        TapTreeIter { leaf_iter: self.0.leaves.iter() }
     }
 }
 
 impl TryFrom<TaprootBuilder> for TapTree {
-    type Error = IncompleteTapTree;
+    type Error = IncompleteBuilder;
 
     /// Constructs [`TapTree`] from a [`TaprootBuilder`] if it is complete binary tree.
     ///
     /// # Returns
-    /// A [`TapTree`] iff the `builder` is complete, otherwise return [`IncompleteTapTree`]
+    /// A [`TapTree`] iff the `builder` is complete, otherwise return [`IncompleteBuilder`]
     /// error with the content of incomplete `builder` instance.
     fn try_from(builder: TaprootBuilder) -> Result<Self, Self::Error> {
-        if !builder.is_finalizable() {
-            Err(IncompleteTapTree::NotFinalized(builder))
-        } else if builder.has_hidden_nodes() {
-            Err(IncompleteTapTree::HiddenParts(builder))
+        builder.try_into_taptree()
+    }
+}
+
+impl TryFrom<NodeInfo> for TapTree {
+    type Error = HiddenNodes;
+
+    /// Constructs [`TapTree`] from a [`NodeInfo`] if it is complete binary tree.
+    ///
+    /// # Returns
+    /// A [`TapTree`] iff the [`NodeInfo`] has no hidden nodes, otherwise return [`HiddenNodes`]
+    /// error with the content of incomplete [`NodeInfo`] instance.
+    fn try_from(node_info: NodeInfo) -> Result<Self, Self::Error> {
+        if node_info.has_hidden_nodes {
+            Err(HiddenNodes::HiddenParts(node_info))
         } else {
-            Ok(TapTree(builder))
+            Ok(TapTree(node_info))
         }
     }
 }
 
 /// Iterator for a taproot script tree, operating in DFS order over leaf depth and
 /// leaf script pairs.
+/// This is guaranteed to not contain any hidden nodes.
+
+// TODO: Enforce this in type system in a later PR. Users should not need to unwrap
+// as script leaf when there are no hidden nodes.
+// An idea can be to have the iterator return a tuple (depth, script, version).
```

Comment:
In 5bf4ca56221dc41488ca8e54aa64351be5ff0471:

I think this TODO is very confusingly worded. I spent some time writing a comment here that you're wrong, but now I understand what you're getting at. I suggest the text
```
// TODO: because a `TapTree` cannot be constructed if it has any hidden nodes, this iterator is
// guaranteed to return script leaves. However, it yields the `LeafNode` type, which may
// represent a hidden node, forcing users to unwrap to get the script. Maybe we should return
// the script (along with depth, version, etc) directly?
```

This is just a nit, but I did spend a bunch of time on it.